File: 第7章 树/二叉搜索树/BinarySearchTree.java
class Node {
	private int data;		// 数据域
	private Node left;		// 左孩子
	private Node right;		// 右孩子
	
	public Node(int val) {
		this.data = val;
		this.left = null;
		this.right = null;
	}

	public Node(int data, Node left, Node right) {
		this.data = data;
		this.left = left;
		this.right = right;
	}
	
	public void setData(int data) {
		this.data = data;
	}

	public int getData() {
		return data;
	}

	public void setLeft(Node left) {
		this.left = left;
	}

	public Node getLeft() {
		return left;
	}

	public void setRight(Node right) {
		this.right = right;
	}

	public Node getRight() {
		return right;
	}
}

public class BinarySearchTree {
	private Node root;		// 树根
	private final static int SPACES = 4;

	/**
	 * @brief  初始化二叉树
	 * @param  val: 根元素
	 */
	public BinarySearchTree(int val) {
		this.root = new Node(val);
	}
	
	public void setRoot(Node root) {
		this.root = root;
	}
	
	public Node getRoot() {
		return this.root;
	}
	
	/**
	 * @brief  先序遍历
	 * @param  node: 树根
	 */
	public void preOrder(Node node) {
		if(node != null) {
			System.out.print(node.getData() + " ");
			preOrder(node.getLeft());
			preOrder(node.getRight());
		}
	}
	
	/**
	 * @brief  中序遍历
	 * @param  node: 树根
	 */
	public void inOrder(Node node) {
		if(node != null) {
			inOrder(node.getLeft());
			System.out.print(node.getData() + " ");
			inOrder(node.getRight());
		}
	}
	
	/**
	 * @brief  后序遍历
	 * @param  node: 树根
	 */
	public void postOrder(Node node) {
		if(node != null) {
			postOrder(node.getLeft());
			postOrder(node.getRight());
			System.out.print(node.getData() + " ");
		}
	}
	
	/**
	 * @brief  打印表示层次的空格符
	 * @note   用于树状打印二叉树
	 * @param  level: 二叉树层次
	 */
	public void indent(int level) {
	    for(int i = 0; i < SPACES * level; i++) {
	        System.out.print(".");
	    }
	}
	
	/**
	 * @brief  树状打印二叉树
	 * @param  node: 树根
	 * @param  level: 二叉树层次
	 */
	void showTree(Node node, int level) {
	    if(node == null) {
	        return;
	    }
	    indent(level);
	    System.out.println(node.getData());
	    level++;
	    showTree(node.getLeft(), level);
	    showTree(node.getRight(), level);
	    level--;
	}
	
	/**
	 * @brief  插入元素
	 * @param  node: 树根
	 * @param  val: 插入值
	 * @retval 操作后树根
	 */
	public Node insert(Node node, int val) {
		if(node == null) {
			return new Node(val);
		}
		
		if(val < node.getData()) {
			node.setLeft(insert(node.getLeft(), val));
		} else {
			node.setRight(insert(node.getRight(), val));
		}
		return node;
	}
	
	/**
	 * @brief  查找元素
	 * @param  node: 树根
	 * @param  val: 查找元素
	 * @retval 元素所在结点指针，不存在返回null
	 */
	public Node search(Node node, int val) {
		if(node == null) {
			return null;
		}
		
		if(val == this.root.getData()) {
			return this.root;
		} else if(val < this.root.getData()) {
			return search(this.root.getLeft(), val);
		} else {
			return search(this.root.getRight(), val);
		}
	}
	
	/**
	 * @brief  获取树高
	 * @param  root: 树根
	 * @retval 树高
	 */
	public int getHeight(Node node) {
		if(node == null) {
			return 0;
		}
		return Math.max(getHeight(node.getLeft()), getHeight(node.getRight())) + 1;
	}

	public static void main(String[] args) {
		BinarySearchTree bst = new BinarySearchTree(5);
		Node root = bst.getRoot();
		
		bst.insert(root, 2);
		bst.insert(root, 7);
		bst.insert(root, 4);
		bst.insert(root, 9);
		bst.insert(root, 1);
		bst.insert(root, 6);
		bst.insert(root, 8);
		bst.insert(root, 3);
		
		System.out.print("preOrder: ");
	    bst.preOrder(root);
	    System.out.println();
	    
	    System.out.print("inOrder: ");
	    bst.inOrder(root);
	    System.out.println();
	    
	    System.out.print("postOrder: ");
	    bst.postOrder(root);
	    System.out.println();

	    bst.showTree(root, 0);
		System.out.println("height: " + bst.getHeight(root));
	}
}

File: Python/第7章 树/二叉搜索树/binary_search_tree.py
class Node:
    """
        结点类
    """
    def __init__(self, data, left=None, right=None):
        """
            初始化结点
            Args:
                data (int): 结点元素
                left (Node): 左孩子
                right (Node): 右孩子
        """
        self.data = data
        self.left = left
        self.right = right

class BinarySearchTree:
    """
        二叉搜索树
    """
    SPACES = 4

    def __init__(self, val):
        """
            初始化二叉树
            Args:
                data (int): 树根元素
        """
        self.root = Node(val)
    
    def pre_order(self, node):
        """
            先序遍历
            Args:
                node (Node): 树根
        """
        if node == None:
            return
        print(node.data, end=' ')
        self.pre_order(node.left)
        self.pre_order(node.right)
    
    def in_order(self, node):
        """
            中序遍历
            Args:
                node (Node): 树根
        """
        if node == None:
            return
        self.in_order(node.left)
        print(node.data, end=' ')
        self.in_order(node.right)
    
    def post_order(self, node):
        """
            后序遍历
            Args:
                node (Node): 树根
        """
        if node == None:
            return
        self.post_order(node.left)
        self.post_order(node.right)
        print(node.data, end=' ')
    
    def indent(self, level):
        """
            打印表示层次的空格符
            Args:
                level (int): 二叉树层次
        """
        for _ in range(self.SPACES * level):
            print(".", end='')
    
    def show_tree(self, node, level):
        """
            树状打印二叉树
            Args:
                node (Node): 树根
                level (int): 二叉树层次
        """
        if node == None:
            return
        self.indent(level)
        print(node.data)
        level += 1
        self.show_tree(node.left, level)
        self.show_tree(node.right, level)
        level -= 1
    
    def insert(self, node, val):
        """
            插入元素
            Args:
                node (Node): 树根
                val (int): 插入值
            Returns:
                [Node]: 操作后树根结点
        """
        # 空树，插入结点设为树根
        if node == None:
            return Node(val)
        
        if val < node.data:
            node.left = self.insert(node.left, val)
        else:
            node.right = self.insert(node.right, val)
        return node
    
    def search(self, node, val):
        """
            查找元素
            Args:
                node (Node): 树根
                val (int): 查找元素
            Returns:
                [Node]: 元素所在结点指针，不存在返回None
        """
        if node == None:
            return None
        
        if val == node.data:
            return node
        elif val < node.data:
            return self.search(node.left, val)
        else:
            return self.search(node.right, val)
    
    def get_height(self, node):
        """
            获取树高
            Args:
                node (Node): 树根
            Returns:
                [int]: 树高
        """
        if node == None:
            return 0
        return max(self.get_height(node.left), self.get_height(node.right)) + 1

def main():
    bst = BinarySearchTree(5)

    bst.insert(bst.root, 2)
    bst.insert(bst.root, 7)
    bst.insert(bst.root, 4)
    bst.insert(bst.root, 9)
    bst.insert(bst.root, 1)
    bst.insert(bst.root, 6)
    bst.insert(bst.root, 8)
    bst.insert(bst.root, 3)

    print("preOrder: ", end='')
    bst.pre_order(bst.root)
    print()

    print("inOrder: ", end='')
    bst.in_order(bst.root)
    print()

    print("postOrder: ", end='')
    bst.post_order(bst.root)
    print()

    bst.show_tree(bst.root, 0)
    print("height: %d" % bst.get_height(bst.root))

if __name__ == "__main__":
    main()